### ISSUE TITLE
Scenario 10 - Heavy CPU Tasks with Cancellation

### ISSUE THREAD

**Issue by jamesward** [jamesward/easyracer#17]
Make 2 Concurrent Requests:
```
GET /10
```
Each will give back a number. Calculate the Fibonacci for each given number, but during calculation, once a second make a request to:
```
GET /10?<num>
```
If the response is a non-2xx reponse, cancel the associated Fibonacci calculation.  Once the Fibonacci number has been calculated (and not cancelled), send a request to get the "right" response:
```
GET /10?<num>=<fibonacci>
```

**Comment by jackgene** [comment]
Just wanted to confirm I understand the purpose of this scenario:
1. The idea is to have the client perform a CPU-bound task (hence the Fibonacci)
2. And perform two of these tasks in parallel (hence the once a second check, if the server sees `GET /10?x`, `GET /10?y`, `GET /10?x`, `GET/10?y`, ... we are good. If the server sees `GET /10?x`, `GET /10?x`, `GET /10?x`, `GET /10?y`, `GET /10?y`, ... client is performing task in series and that's not good.

My only concern is that because of the memoizable nature of Fibonacci, it's easy for clients to "game the system". In fact, you could argue that we are intentionally doing the sub-optimal thing, by performing the calculations in parallel, duplicating a bunch of calculations.

In Scala for instance, this single threaded approach of calculating two Fibonacci's should perform similarly with a parallel solution, while using only one core:
```scala
  def fibonacci(n1: Int, n2: Int): (BigInt, BigInt) = {
    @tailrec
    def fibonacciTail(n: Int, a: BigInt, b: BigInt): (BigInt, BigInt) = {
      if (n == 0) (0, a)
      else if (n == 1) (a, b)
      else fibonacciTail(n - 1, b, a + b)
    }

    val nSmaller = math.min(n1, n2)
    val nLarger = math.max(n1, n2)
    val (fibSmallerMinus1, fibSmaller) = fibonacciTail(nSmaller, 0, 1)
    val (_, fibLarger) = fibonacciTail(nLarger - nSmaller + 1, fibSmallerMinus1, fibSmaller)

    if (n1 > n2) (fibLarger, fibSmaller) else (fibSmaller, fibLarger)
  }
```

I generally think this is fine, and I don't think people are here to "game the system" anyway, but just thought I'd share this concern.

In looking into this, I had a look at https://en.wikipedia.org/wiki/Embarrassingly_parallel to see if there may be algorithms that are more suited for this than Fibonacci, and one that jumped out is perhaps a simplified "proof of work" task.

**Comment by jackgene** [comment]
Was thinking about it more, and realized an "embarrassingly parallel" solution is perhaps not what we want, but rather an unparallelizable one (which Fibonacci is). My concern is more about how easy Fibonacci is to optimize because of its memoizable nature. But I think as long as it's made clear that the point is not to memoize, we should be good.

**Comment by jackgene** [comment]
Another thought, as I was thinking about how this would be implemented in the various language, I realize that making Fibonacci (or any CPU-bound function) cancellable is in and of itself an interesting topic:
- It can be trivial to implement in languages like Go (where you'd just check the `context` on every iteration to see if it's cancelled) or Swift (similar, but using a `task.yield()` I think - still need to investigate).
- Or it can be a completely different implementation in Elm (Fibonacci without cancellation is just a function, with cancellation, you need to include The Elm Architecture, which makes it a combinations of functions, messages, etc).

So I'm thinking it's a good idea to have a scenario where you are just required to implement Fibonacci without cancellation. WDYT?

**Comment by jackgene** [comment]
Here's comparing Go regular vs cancellable Fibonacci:
```go
func Fibonacci(n int) *big.Int {
	fib, lastFib := big.NewInt(0), big.NewInt(1)
	if n < 1 {
		return fib
	}
	for i := 0; i < n; i++ {
		lastFib.Add(lastFib, fib)
		lastFib, fib = fib, lastFib
	}

	return fib
}

func CancellableFibonacci(ctx context.Context, n int) (*big.Int, error) {
	fib, lastFib := big.NewInt(0), big.NewInt(1)
	if n < 1 {
		return fib, nil
	}
	yield := make(chan struct{}, 1)
	for i := 0; i < n; i++ {
		lastFib.Add(lastFib, fib)
		lastFib, fib = fib, lastFib
		yield <- struct{}{}
		select {
		case <-ctx.Done():
			return nil, ctx.Err()
		case <-yield:
		}
	}

	return fib, nil
}
```

**Comment by jackgene** [comment]
And Swift:
```swift
func fibonacci(_ n: Int) -> BInt {
    var fib = BInt(0), lastFib = BInt(1)
    
    if n < 1 {
        return fib
    }
    for _ in 1...n {
        lastFib += fib
        swap(&fib, &lastFib)
    }
    return fib
}
```

To make cancellable, make the function `throws`, and `Task.checkCancellation()` on every iteration:
```swift
func cancellableFibonacci(_ n: Int) throws -> BInt {
    var fib = BInt(0), lastFib = BInt(1)
    
    if n < 1 {
        return fib
    }
    for _ in 1...n {
        lastFib += fib
        swap(&fib, &lastFib)
        try Task.checkCancellation()
    }
    return fib
}
```

**Comment by jamesward** [comment]
Yeah Jack. You got it. Point will be to show how hard it is to cancel a tight loop / cpu bound task. Agree that we would have to clarify that optimizing with memoization is outside of what we are trying to explore. I'll start working on the server-side of this and will see how it evolves when building a real client.

**Comment by jamesward** [comment]
So a problem with this scenario idea is that we don't know how long it should take to calculate a given Fibonacci so we don't know how much time we should wait to respond with the cancellation command.  We could just respond on the first one, but that kinda defeats some of the purpose.

Another issue is that there isn't any way to validate that cancellation actually happens from the server-side.

Maybe instead there would be some way for the cancel checker to also report it's progress.

Still not sure if there is any way to validate the cancellation though.

**Comment by jackgene** [comment]
Yeah, those are definitely challenges.

For #1, I've seen huge variances in runtime for the same naïve Fibonacci implementation, calculating `fibonacci(1_000_000)`, from 3 seconds (Go), to 20 seconds (Swift), to "unclear if it'll complete, but I got tired of waiting" (Elm). One potential solution is to start the scenario with a baseline run of calculating the fibonacci of around 10,000 to determine runtime performance. And then generate numbers based on that baseline?

For #2, I don't think it's possible to ever validate that the cancellation actually happened. I like the cancel checker reporting progress idea, but perhaps makes the implementation a bit complex?

An alternative to cancel checker reporting progress, is for the client to acknowledge cancellation by responding with the iteration it's on, at the time of cancellation, and the fibonacci for that iteration. We won't be able to validate that the task was cancelled, but at least this demonstrates that "things are happening" between steps, and hopefully shows that, whether or not the task is cancelled, the implementation likely is cancellable?

Like, "if I can report my state mid-calculation, I can also stop what I'm doing." But I'm not sure how true that is.

**Comment by jamesward** [comment]
Here are some new thoughts on this...

**Scenario 10**
Goal: Be able to complete two parallel tasks, faster than sequentially

Assuming a recursive SHA-512 function that performs repeated hashing for >5 seconds where each iteration appends the hash to the new input, like:
```scala
  @tailrec
  def shaLoop(count: Int, maxTimeNanos: Long, contents: Array[Byte], messageDigest: MessageDigest): (Int, Array[Byte]) =
    if java.lang.System.nanoTime() >= maxTimeNanos then
      (count, contents)
    else
      val newContents = messageDigest.digest(contents)
      shaLoop(count + 1, maxTimeNanos, contents ++ newContents, messageDigest)
```

1. Make a request to `/10` to get the initial 64 bytes to hash from the response body.
1. Make a request to `/10?<64 bytes>` and write the results into the request body to establish a baseline. The response body will include 2 lines which are the initial 64 bytes to hash in parallel
1. Make a request to `/10?<64 bytes><64 bytes>` and write the results into the request body (in the corresponding order). If heuristically the tasks were faster in parallel (with some margin) then you will get the right answer.

**Scenario 11**
Goal: Be able to cancel a CPU intensive task

For this scenario you will need a version of the SHA-512 function which takes >5 seconds to run but can also be cancelled earlier.

1. Make a request to `/11` to get two lines from the response body containing the initial 64 bytes to hash.
1. Make two concurrent requests to `/11?<64 bytes>` (one for each of the initial values from the step 1) and stream the results into the request body, but one of the requests should stop writing results after 2 seconds (using cancellation, not a max time limit) and end the request body.  Once both requests complete, combine their responses in order of arrival to form the "right" answer.

Notes: I might need to think more about game prevention on the cancellation part.  Also, I'm not sure if the streaming the results to the request body is necessary to show the cancellation part of this.

**Comment by jamesward** [comment]
I've been noodling this and have another idea...

**Scenario 10**

Use a similar recursive sha512 loop as above but also a sha256 one.
1. Make a request to `/10?<64 bytes>`
2. Begin a race that calculates the recursive sha256 & sha512 values for the provided 64 bytes in parallel
3. When the first one reaches x iterations, complete it causing it to win the race and the other should then be cancelled (Note that I'm not sure what x should be as it could depend heavily on the platform and how quick it can calculate.  I've thought about coming up with a way to baseline or making things time based but they all seem to have undesired tradeoffs.  So my current thinking is that once we get some different platform impls of this we can see how significantly x varies if we aim for about 2-5 seconds to get a winner. Note that if the server determines the x in some way, that could be used for verifying what the user calculated.)
4. Put the result in the request body (Note that I'm not sure how hard async request body writes & reads are in any of the technologies currently used with EasyRacer)
5. If you've done things correctly, you'll get the "right" response

It looks like assuming the client & server are running on the same machine (which I think should nearly always be the case for this stuff) the server can actually watch the host CPU usage via `/proc/stat` and verify that two CPUs were (with some margin) fully saturated for the duration of the request and that they both stop saturating the CPU at the same time (with some margin).  A few things that we'd likely need to investigate: for dev on Windows & Mac does `/proc/stat` work to show host data?  Right now the scenario-server uses a `scratch` base image and we'd have to change that.  Another possible option would be to have a "background request" in this scenario that would feed CPU data to the server every 100ms.

Another possible issue with this whole idea is that some languages & platforms may not be able to actually do this (assuming we won't take crazy escape hatches).  So how do we convey that?  Maybe the README should report something like 9/10 (or some progress like thing).

Ultimately what we are trying to show is how structured concurrency relates to cancellation & parallelism with blocking tasks.  So far this is the easiest way I've thought of to capture that but it has some gaps.

**Comment by jamesward** [comment]
Here is another approach I'm thinking about:

1. Make a GET request to `/10` and while the connection is open, perform something computationally heavy (doesn't really matter what)
2. Parallel to 1, every 100ms, send the current process load (0 to 1) to `/10?load={value}`
3. After the connection in 1 closes, cancel the computationally heavy task, wait 100ms and send a final request with the process load to `/10?load={value}`

You'll get the "right" response if the load is near 1 for the duration of the open connection and then drops to near zero after the connection closes.

On the JVM process load can be fetched with:
```
import java.lang.management.ManagementFactory
import com.sun.management.OperatingSystemMXBean

val osBean = ManagementFactory.getPlatformMXBean(classOf[OperatingSystemMXBean])
val load = osBean.getProcessCpuLoad * osBean.getAvailableProcessors
```

I think this will achieve the goal of showing how computationally heavy tasks can be cancelled and how they can be done in parallel with other tasks.  It'll be interesting to see how different languages / frameworks handle this.  In Scala with ZIO it is pretty straightforward, something like:
```
val messageDigest = MessageDigest.getInstance("SHA-512")
val seed = Random.nextBytes(512).run

val blocking = ZIO.attemptBlockingInterrupt:
  var result = seed.toArray
  while (!Thread.interrupted())
    result = messageDigest.digest(result)

  val reporter = ZIO.attempt:
    val osBean = ManagementFactory.getPlatformMXBean(classOf[OperatingSystemMXBean])
    println(osBean.getProcessCpuLoad * osBean.getAvailableProcessors)

  reporter.repeat(Schedule.spaced(100.millis).upTo(5.seconds)).race(blocking).run
```

It is cool that the `race` can be used here too (note that this doesn't do the final part yet of sending the load 100ms after the race completes but that part should be easy).

**Comment by jamesward** [comment]
Here is the working client:
https://github.com/jamesward/easyracer/blob/main/scala-zio/src/main/scala/EasyRacerClient.scala#L108-L134

The scenario changed a bit:
1. Make a GET request to `/10` and while the connection is open, perform something computationally heavy (doesn't really matter what), cancelling that task when the connection closes.
2. Parallel to 1, every 1 second, send the current process load (0 to 1) to `/10?load={value}`, which will respond with a 20x response if it looks like 1 was accomplished (in which case you can stop sending load values), otherwise it will respond with a 10x response if you should continue sending values, or with a 40x response if something has gone wrong.

**Comment by jamesward** [comment]
Small change to the endpoint, to make the lifecycle of the server-side work better, the endpoint now needs an "id" for the requests, so: `/10?{some_id}` for the blocker and `/10?{some_id}={load}` for the monitor. Sequential runs should use a different id.  There is also now a number of impls:
https://github.com/jamesward/easyracer/blob/main/java-loom/src/main/java/Main.java#L229-L284
https://github.com/jamesward/easyracer/blob/main/kotlin-coroutines/src/main/kotlin/Main.kt#L177-L218
https://github.com/jamesward/easyracer/blob/main/scala-ox/src/main/scala/EasyRacerClient.scala#L83-L114
https://github.com/jamesward/easyracer/blob/main/scala-zio/src/main/scala/EasyRacerClient.scala#L108-L137

**Comment by jamesward** [comment]
I've added this scenario to the main README
